sdk/lib/drivers/arbiter/arbiter.c

```diff
@@ -34,10 +230,135 @@ ArbInitializeArbiterInstance(
 {
     NTSTATUS Status;
 
-    DPRINT("ArbInitializeArbiterInstance: Initializing %S Arbiter\n", ArbiterName);
-    UNIMPLEMENTED;
+    DPRINT("ArbInitializeArbiterInstance: '%S'\n", ArbiterName);
+
+    ASSERT(Arbiter->UnpackRequirement != NULL);
+    ASSERT(Arbiter->PackResource != NULL);
+    ASSERT(Arbiter->UnpackResource != NULL);
+
+    ASSERT(Arbiter->MutexEvent == NULL &&
+           Arbiter->Allocation == NULL &&
+           Arbiter->PossibleAllocation == NULL &&
+           Arbiter->AllocationStack == NULL);
+
+    Arbiter->Signature = 'sbrA';
+    Arbiter->BusDeviceObject = BusDeviceObject;
+
+    Arbiter->MutexEvent = ExAllocatePoolWithTag(NonPagedPool, sizeof(KEVENT), 'MbrA');
+    if (!Arbiter->MutexEvent)
+    {
+        DPRINT1("ArbInitializeArbiterInstance: STATUS_INSUFFICIENT_RESOURCES\n");
+        return STATUS_INSUFFICIENT_RESOURCES;
+    }
+
+    KeInitializeEvent(Arbiter->MutexEvent, SynchronizationEvent, TRUE);
+
+    Arbiter->AllocationStack = ExAllocatePoolWithTag(PagedPool, PAGE_SIZE, 'AbrA');
+    if (!Arbiter->AllocationStack)
+    {
+        DPRINT1("ArbInitializeArbiterInstance: STATUS_INSUFFICIENT_RESOURCES\n");
+        ExFreePoolWithTag(Arbiter->MutexEvent, 'MbrA');
+        return STATUS_INSUFFICIENT_RESOURCES;
+    }
+
+    Arbiter->AllocationStackMaxSize = PAGE_SIZE;
+
+    Arbiter->Allocation = ExAllocatePoolWithTag(PagedPool, sizeof(RTL_RANGE_LIST), 'RbrA');
+    if (!Arbiter->Allocation)
+    {
+        DPRINT1("ArbInitializeArbiterInstance: STATUS_INSUFFICIENT_RESOURCES\n");
+        ExFreePoolWithTag(Arbiter->AllocationStack, 'AbrA');
+        ExFreePoolWithTag(Arbiter->MutexEvent, 'MbrA');
+        return STATUS_INSUFFICIENT_RESOURCES;
+    }
+
+    Arbiter->PossibleAllocation = ExAllocatePoolWithTag(PagedPool, sizeof(RTL_RANGE_LIST), 'RbrA');
+    if (!Arbiter->PossibleAllocation)
+    {
+        DPRINT1("ArbInitializeArbiterInstance: STATUS_INSUFFICIENT_RESOURCES\n");
+        ExFreePoolWithTag(Arbiter->Allocation, 'RbrA');
+        ExFreePoolWithTag(Arbiter->AllocationStack, 'AbrA');
+        ExFreePoolWithTag(Arbiter->MutexEvent, 'MbrA');
+        return STATUS_INSUFFICIENT_RESOURCES;
+    }
+
+    RtlInitializeRangeList(Arbiter->Allocation);
+    RtlInitializeRangeList(Arbiter->PossibleAllocation);
+
+    Arbiter->Name = ArbiterName;
+    Arbiter->ResourceType = ResourceType;
+    Arbiter->TransactionInProgress = FALSE;
+
+    if (!Arbiter->TestAllocation)
```

Comment:
Why checking these? Isn't this function supposed to be called once for each arbiter instance?

Reply:
Assigned globally here if not assigned in an instance of an arbiter.